src/AI/WallyAI.java

```diff
@@ -0,0 +1,788 @@
+              {
+                log(String.format("    is it safe to go to %s?", xyc));
+                if( canWallHere(gameMap, threatMap, unit, xyc) )
+                {
+                  log(String.format("    Yes"));
+                  destination = xyc;
+                  break;
+                }
+              }
+              if( null != destination )
+              {
+                Path movePath = Utils.findShortestPath(unit, destination, gameMap);
+                if( movePath.getPathLength() > 1 ) // We only want to try to travel if we can actually go somewhere
+                {
+                  GameAction move = new GameAction.WaitAction(unit, movePath);
+                  actions.offer(move);
+                  stateChange = true;
+                  break;
+                }
+              }
+            }
+          }
+        }
+      }
+
+      // Check for an available buying enhancement power
+      if( actions.isEmpty() && !stateChange )
+      {
+        AIUtils.queueCromulentAbility(actions, myCo, CommanderAbility.PHASE_BUY);
+      }
+
+      // We will add all build commands at once, since they can't conflict.
+      if( actions.isEmpty() && !stateChange )
+      {
+        queueUnitProductionActions(gameMap);
+      }
+
+      // Check for a turn-ending power
+      if( actions.isEmpty() && !stateChange )
+      {
+        AIUtils.queueCromulentAbility(actions, myCo, CommanderAbility.PHASE_TURN_END);
+      }
+
+      // Return the next action, or null if actions is empty.
+      nextAction = actions.poll();
+    } while (nextAction == null && stateChange); // we don't want to end early, so if the state changed and we don't have an action yet, try again
+    log(String.format("  Action: %s", nextAction));
+    return nextAction;
+  }
+
+  private boolean isSafe(GameMap gameMap, Map<UnitModel, Map<XYCoord, Double>> threatMap, Unit unit, XYCoord xyc)
+  {
+    Double threat = threatMap.get(myCo.getUnitModel(unit.model.type)).get(xyc);
+    int threshhold = unit.model.hasDirectFireWeapon() ? DIRECT_THREAT_THRESHHOLD : INDIRECT_THREAT_THRESHHOLD;
+    return (null == threat || threshhold > threat);
+  }
+
+  /**
+   * @return whether it's safe or a good place to wall
+   */
+  private boolean canWallHere(GameMap gameMap, Map<UnitModel, Map<XYCoord, Double>> threatMap, Unit unit, XYCoord xyc)
+  {
+    // if we're safe, we're safe
+    if( isSafe(gameMap, threatMap, unit, xyc) )
+      return true;
+    // don't stand on a factory or in danger for no good reason
+    if( gameMap.getEnvironment(xyc).terrainType == TerrainType.FACTORY || !unit.model.hasDirectFireWeapon())
+      return false;
+
+    // TODO: Determine whether the ally actually needs a wall there. Mechs walling for Tanks vs inf is... silly.
+    // if we'd be a nice wall for a worthy ally, we can pretend we're safe there also
+    XYCoord[] adjacentCoords = { new XYCoord(xyc.xCoord + 1, xyc.yCoord), new XYCoord(xyc.xCoord - 1, xyc.yCoord),
```

Comment:
This might be a good candidate for `Utils.findLocationsInRange()`.

Reply:
Yeah, that's fair.